core/trino-main/src/main/java/io/trino/sql/planner/optimizations/DeterminePreferredDynamicFilterTimeout.java

```diff
@@ -0,0 +1,317 @@
+                return dynamicFilterSourceNode.getSource();
+            }
+            throw new IllegalArgumentException("Plan node unsupported " + planNode.getClass().getSimpleName());
+        }
+
+        private DynamicFilterTimeout getBuildSideState(PlanNode planNode, Symbol dynamicFilterSymbol)
+        {
+            // Skip for expanding plan nodes like CROSS JOIN or UNNEST which can substantially increase the amount of data.
+            if (isInputMultiplyingPlanNodePresent(planNode)) {
+                return DynamicFilterTimeout.NO_WAIT;
+            }
+
+            SymbolStatsEstimate symbolStatsEstimate = statsProvider.getStats(planNode).getSymbolStatistics(dynamicFilterSymbol);
+            if (!symbolStatsEstimate.isUnknown() && !isExpandingPlanNodePresent(planNode)) {
+                if (symbolStatsEstimate.getDistinctValuesCount() < awaitedDynamicFilterMaxNdvCount) {
+                    return DynamicFilterTimeout.USE_PREFERRED_TIMEOUT;
+                }
+            }
+
+            double rowCount = getEstimatedMaxOutputRowCount(planNode, statsProvider);
+            if (isNaN(rowCount)) {
+                return DynamicFilterTimeout.UNESTIMATED;
+            }
+            if (rowCount < awaitedDynamicFilterMaxRowCount) {
+                return DynamicFilterTimeout.USE_PREFERRED_TIMEOUT;
+            }
+
+            return DynamicFilterTimeout.NO_WAIT;
+        }
+    }
+
+    private static boolean isInputMultiplyingPlanNodePresent(PlanNode root)
+    {
+        return PlanNodeSearcher.searchFrom(root)
+                .where(DeterminePreferredDynamicFilterTimeout::isInputMultiplyingPlanNode)
+                .matches();
+    }
+
+    private static boolean isInputMultiplyingPlanNode(PlanNode node)
+    {
+        if (node instanceof UnnestNode) {
+            return true;
+        }
+
+        if (node instanceof JoinNode joinNode) {
+            // Skip for cross join
+            if (joinNode.isCrossJoin()) {
+                // If any of the input node is scalar then there's no need to skip cross join
+                return !isAtMostScalar(joinNode.getRight()) && !isAtMostScalar(joinNode.getLeft());
+            }
+
+            // Skip for joins with multi keys since output row count stats estimation can wrong due to
+            // low correlation between multiple join keys.
+            return joinNode.getCriteria().size() > 1;
+        }
+
+        return false;
+    }
+
+    private static boolean isExpandingPlanNodePresent(PlanNode root)
+    {
+        return PlanNodeSearcher.searchFrom(root)
+                .where(DeterminePreferredDynamicFilterTimeout::isExpandingPlanNode)
+                .matches();
+    }
+
+    private static boolean isExpandingPlanNode(PlanNode node)
+    {
+        return node instanceof JoinNode
+               // consider union node and exchange node with multiple sources as expanding since it merge the rows
```

Comment:
should be more like `union and exchange nodes that consume from multiple source are considered expanding since they produce data than their children individually`